server/ocsp.go

```diff
@@ -0,0 +1,377 @@
+
+	reqEnc := base64.StdEncoding.EncodeToString(reqDER)
+
+	responders := oc.Leaf.OCSPServer
+	if len(overrideURLs) > 0 {
+		responders = overrideURLs
+	}
+
+	oc.mu.Lock()
+	hc := oc.hc
+	oc.mu.Unlock()
+	var raw []byte
+	for _, u := range responders {
+		u = strings.TrimSuffix(u, "/")
+		raw, err = getRequestBytes(fmt.Sprintf("%s/%s", u, reqEnc), hc)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return nil, nil, fmt.Errorf("exhausted ocsp servers: %w", err)
+	}
+
+	resp, err := ocsp.ParseResponse(raw, oc.Issuer)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := validOCSPResponse(resp); err != nil {
+		return nil, nil, err
+	}
+
+	oc.mu.Lock()
+	oc.raw = raw
+	oc.resp = resp
+	oc.mu.Unlock()
+
+	return raw, resp, nil
+}
+
+func (oc *OCSPMonitor) run() {
+	s := oc.srv
+	s.mu.Lock()
+	quitCh := s.quitCh
+	s.mu.Unlock()
+
+	defer s.grWG.Done()
+	config := s.getOpts().OCSPConfig
+	shouldShutdown := config != nil && (config.Mode == OCSPModeMust || config.Mode == OCSPModeAlways)
+
+	var nextRun time.Duration
+	_, resp, err := oc.getStatus()
+	if err == nil && resp.Status == ocsp.Good {
+		nextRun = oc.getNextRun()
+		t := resp.NextUpdate.Format(time.RFC3339Nano)
+		s.Noticef(
+			"Found existing OCSP certificate status: good, next update %s, checking again in %s",
+			t, nextRun,
+		)
+	} else if err == nil && shouldShutdown {
+		// If resp.Status is ocsp.Revoked, ocsp.Unknown, or any other value.
+		s.Errorf("Found existing OCSP certificate status: %s", ocspStatusString(resp.Status))
+		s.Shutdown()
+		return
+	}
+
+	for s.Running() {
+		select {
+		case <-time.After(nextRun):
+		case <-quitCh:
+			return
+		}
+		_, resp, err := oc.getRemoteStatus()
+		if err != nil {
+			nextRun = oc.getNextRun()
+			s.Errorf("Bad OCSP status update: %s, trying again in %v", err, nextRun)
+			continue
+		}
+
+		switch n := resp.Status; n {
+		case ocsp.Good:
+			nextRun = oc.getNextRun()
+			t := resp.NextUpdate.Format(time.RFC3339Nano)
+			s.Noticef(
+				"Received OCSP certificate status: good, next update %s, checking again in %s",
+				t, nextRun,
+			)
+			continue
+		default:
+			s.Errorf("Received OCSP certificate status: %s, shutting down", ocspStatusString(n))
+			s.Shutdown()
+			return
+		}
+	}
+}
+
+// NewOCSPMonitor takes a TLS configuration then wraps it with the callbacks set for OCSP verification
+// along with a monitor that will periodically fetch OCSP staples.
+func (srv *Server) NewOCSPMonitor(tc *tls.Config) (*tls.Config, *OCSPMonitor, error) {
+	opts := srv.getOpts()
+	oc := opts.OCSPConfig
+	tcOpts := opts.TLSConfigOpts
+	if oc.Mode == OCSPModeNever {
+		return tc, nil, nil
+	}
+	cert, err := tls.LoadX509KeyPair(tcOpts.CertFile, tcOpts.KeyFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	cert.Leaf, err = x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	if oc.Mode == OCSPModeAuto && !hasOCSPStatusRequest(cert.Leaf) {
+		// "status_request" flag not set. No need to do anything.
```

Comment:
Is NewOCSPMonitor() called again on reload when a new certificate is seen, or does this mean that if we start with a certificate without must-staple and later replace it with a must-staple cert then we won't set up the OCSP staples for it?

Reply:
Yes we don't handle reload in this PR, but that is what should be done.

Reply:
It is reload aware now, monitors are restarted when TLS certs change.